lib/spack/spack/build_systems/cmake.py

```diff
@@ -372,7 +372,14 @@ def build_dirname(self):
     @property
     def build_directory(self):
         """Full-path to the directory to use when building the package."""
-        return os.path.join(self.pkg.stage.path, self.build_dirname)
+        stage_build_path = self.pkg.stage.build_directory
+        # default build stage is in source, if this is the case
+        # establish in stage build dir. If stage returns a non
```

Comment:
This comment is confusing:

* This method is supposed to return a directory that enables out-of-source builds
* When using an external build directory root, we achieve this naturally
* If not, then we place the external build directory in the stage

Possible clarifications:

* when you refer to a "default" build stage, do you mean the default chosen by CMake? also I think by "build stage" you mean "build directory"

Reply:
I think the refactor clarified this behavior to the point where we don't need too much of an elaboration, so I've removed much of this comment. Will push with rest of changes.